### PR TITLE
ES6: migrated "util"

### PR DESCRIPTION
**Pull request by bennyn**
`PromiseQueue.coffee` will not be part of this conversion as per request from @gregor (he will convert it later).

### CODE REVIEW

File: app/script/main/app.coffee [hunk 1]
```diff
@@ -239,7 +239,7 @@ class z.main.App
       if is_reload and error.type not in [z.client.ClientError.TYPE.MISSING_ON_BACKEND, z.client.ClientError.TYPE.NO_LOCAL_CLIENT]
         @auth.client.execute_on_connectivity(z.service.Client::CONNECTIVITY_CHECK_TRIGGER.APP_INIT_RELOAD).then -> window.location.reload false
       else if navigator.onLine
-        @logger.error "Caused by: #{error?.message or error}"
+        @logger.error "Caused by: #{error?.message or error}", error
```
Comment (gregor) [inline app/script/main/app.coffee]: Problematic. If error is a string as it sometimes is and not the error object, we will end up with a log message that does no longer contain the "Caused by reference". Maybe we should be more explicit and do a proper _.isObject check with tow error logging cases.

File: app/script/util/ArrayUtil.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    // item is last item in the array
+    if ((next_index === array.length) && (index > 0)) {
+      return array[index - 1];
+    }
+
+    if (next_index >= array.length) {
+      return undefined;
+    }
+
+    for (let i = next_index; i <= array.length; i++) {
```
Comment (gregor) [inline app/script/util/ArrayUtil.js:63]: You have a proper array. We can use that:

````
for (const item of array) {
  if(!_.isFunction(filter) || !!filter(item)) {
    return item
  }
}
````

File: app/script/util/Crypto.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+window.z.util = z.util || {};
+
+z.util.Crypto = {
+  Hashing: {
+    joaat_hash(str) {
+      const {uint32} = window;
+      let hash = uint32.toUint32(0);
+      const key = str.toLowerCase();
+
+      for (let i = 0; i <= key.length - 1; i++) {
```
Comment (gregor) [inline app/script/util/Crypto.js:32]:
````
for (const index of [0..key.length - 1])
````

File: app/script/util/ArrayUtil.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+   * @returns {Array<any>} new array with interpolated values
+   */
+  interpolate(array, length) {
+    const new_array = [];
+    const scale_factor = (array.length - 1) / (length - 1);
+
+    new_array[0] = array[0];
+    new_array[length - 1] = array[array.length - 1];
+
+    for (let i = 1; i < length - 1; i++) {
```
Comment (gregor) [inline app/script/util/ArrayUtil.js:84]: `for (const index of [0..length  - 1])`

File: app/script/util/DebugUtil.js [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+window.z = window.z || {};
+window.z.util = z.util || {};
+
+z.util.DebugUtil = class DebugUtil {
+  constructor(user_repository, conversation_repository) {
+    this.conversation_repository = conversation_repository;
+    this.logger = new z.util.Logger('z.util.DebugUtil', z.config.LOGGER.OPTIONS);
+    this.user_repository = user_repository;
+
+    this.get_event_info = this.get_event_info.bind(this);
```
Comment (gregor) [inline app/script/util/DebugUtil.js]: No need for these

File: app/script/util/DebugUtil.js [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    this.get_notification_from_stream = this.get_notification_from_stream.bind(this);
+    this.get_notifications_from_stream = this.get_notifications_from_stream.bind(this);
+    this.get_number_of_clients_in_conversation = this.get_number_of_clients_in_conversation.bind(this);
+    this.get_v2_call_participants = this.get_v2_call_participants.bind(this);
+  }
+
+  block_all_connections() {
+    const block_users = [];
```
Comment (gregor) [inline app/script/util/DebugUtil.js]:
````
return Promise.all(wire.app.repository.user.users().map((user_et) => {
  return this.user_repository.block_user(user_et);
}));
````
or 

````
const block_users = wire.app.repository.user.users().map((user_et) => {
  return this.user_repository.block_user(user_et);
});

return Promise.all(block_users);
````

File: app/script/util/DebugUtil.js [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+      if (matching_notifications.length) {
+        return matching_notifications[0];
+      }
+
+      if (has_more) {
+        const last_notification = notifications[notifications.length - 1];
+        return this.get_notification_from_stream(notification_id, last_notification.id);
+      }
+      return this.logger.log(`Notification '${notification_id}' was not found in encrypted notification stream`);
+
```
Comment (gregor) [inline app/script/util/DebugUtil.js]: Weird empty line

File: app/script/util/DebugUtil.js [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+      }
+      );
+  }
+
+  get_info_for_client_decryption_errors(remote_user_id, remote_client_id) {
+    return Promise.all([
+      this.get_notifications_from_stream(remote_user_id, remote_client_id),
+      this.get_serialised_identity(),
+      this.get_serialised_session(`${remote_user_id}@${remote_client_id}`),
+    ])
```
Comment (gregor) [inline app/script/util/DebugUtil.js:183]: Indentation

File: app/script/util/DebugUtil.js [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    return wire.app.service.notification.get_notifications(client_scope, notification_id_since, 10000)
+      .then(_got_notifications);
+  }
+
+  get_objects_for_decryption_errors(session_id, notification_id) {
+    return Promise.all([
+      this.get_notification_from_stream(notification_id),
+      this.get_serialised_identity(),
+      this.get_serialised_session(session_id),
+    ])
```
Comment (gregor) [inline app/script/util/DebugUtil.js:168]: Indentation

File: app/script/util/DebugUtil.js [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+      .then((resolve_array) => {
+        return JSON.stringify({
+          identity: resolve_array[1],
+          notification: resolve_array[0],
+          session: resolve_array[2],
+        });
+      }
+      );
```
Comment (gregor) [inline app/script/util/DebugUtil.js]: Combine with line above

File: app/script/util/LocalizerUtil.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+'use strict';
+
+window.z = window.z || {};
+window.z.util = z.util || {};
+
+z.util.LocalizerUtil = {
+  join_names: function(user_ets, declension = z.string.Declension.ACCUSATIVE) {
+    const names_string = user_ets.map((user_et) => z.util.get_first_name(user_et, declension)).join(', ');
```
Comment (gregor) [inline app/script/util/LocalizerUtil.js]: Pretty long line, can we format this with {} for easier readabilty

File: app/script/util/StorageUtil.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+z.util.StorageUtil = {
+  get_value: function(key) {
+    return amplify.store(key);
+  },
+  reset_value: function(key) {
+    return z.util.StorageUtil.set_value(key, null);
+  },
+  set_value: function(key, value, seconds_to_expire) {
+    if (seconds_to_expire) {
+      amplify.store(key, value, {expires: seconds_to_expire * 1000});
```
Comment (gregor) [inline app/script/util/StorageUtil.js]: I think you have to add returns here as well as some parts of the code expect the stored value to be immediately returned.

````
if (seconds_to_expire) {
  return amplify.store(key, value, {expires: seconds_to_expire * 1000});
}
return amplify.store(key, value);
````

File: app/script/util/protobuf.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  load_protos: function(file) {
+    return new Promise(function(resolve, reject) {
+      return dcodeIO.ProtoBuf.loadProtoFile(file, function(error, builder) {
+        if (error) {
+          return reject(new Error(`Loading protocol buffer file failed: ${error.message}`));
+        } 
+        z.proto = z.proto || {};
+        _.extend(z.proto, builder.build());
+        return resolve();
+
```
Comment (gregor) [inline app/script/util/protobuf.js]: Weird empty line

File: app/script/util/util.js [hunk 14]
```diff
@@ -0,0 +1,703 @@
...
+z.util.load_image = function(blob) {
+  return new Promise(function(resolve, reject) {
+    const object_url = window.URL.createObjectURL(blob);
+    const img = new Image();
+    img.onload = function() {
+      resolve(this);
+      return window.URL.revokeObjectURL(object_url);
+    };
+    img.onerror = reject;
+    return img.src = object_url;
```
Comment (gregor) [inline app/script/util/util.js]: No return

File: app/script/util/util.js [hunk 15]
```diff
@@ -0,0 +1,703 @@
...
+};
+
+z.util.load_file_buffer = function(file) {
+  return new Promise(function(resolve, reject) {
+    const reader = new FileReader();
+    reader.onload = function() {
+      return resolve(this.result);
+    };
+    reader.onerror = reject;
+    return reader.readAsArrayBuffer(file);
```
Comment (gregor) [inline app/script/util/util.js]: No return

File: app/script/util/util.js [hunk 16]
```diff
@@ -0,0 +1,703 @@
...
+        return resolve([xhr.response, xhr.getResponseHeader('content-type')]);
+      }
+      return reject(new Error(`Requesting arraybuffer failed with status ${xhr.status}`));
+
+    };
+    xhr.onerror = reject;
+    if (typeof xhr_accessor_function === 'function') {
+      xhr_accessor_function(xhr);
+    }
+    return xhr.send();
```
Comment (gregor) [inline app/script/util/util.js]: No return

File: app/script/util/util.js [hunk 17]
```diff
@@ -0,0 +1,703 @@
...
+  return new Promise(function(resolve, reject) {
+    const xhr = new XMLHttpRequest();
+    xhr.open('GET', url, true);
+    xhr.responseType = 'arraybuffer';
+    xhr.onload = function() {
+      if (xhr.status === 200) {
+        return resolve([xhr.response, xhr.getResponseHeader('content-type')]);
+      }
+      return reject(new Error(`Requesting arraybuffer failed with status ${xhr.status}`));
+
```
Comment (gregor) [inline app/script/util/util.js]: Weird empty line

File: app/script/util/worker.js [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+window.z.util = z.util || {};
+
+z.util.Worker = class Worker {
+  constructor(uri) {
+    this.post = (data) => {
+      return new Promise(function(resolve, reject) {
+        const worker = new window.Worker(uri);
+        worker.onmessage = (event) => resolve(event.data);
+        worker.onerror = (error) => reject(error);
+        return worker.postMessage(data);
```
Comment (gregor) [inline app/script/util/worker.js]: No return

File: app/script/main/app.coffee (hunk 1 above)
Comment (bennycode) [reply app/script/main/app.coffee]: Hm, this is due to our logger implementation, right?

**Pull request merged**